### PR TITLE
detect circular reference for specs

### PR DESCRIPTION
**Pull request by ruowan**

### CODE REVIEW

File: src/graph.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+const getReferencedFileNames = (fileName: string, doc: json.Json) => {
+  const dir = path.dirname(fileName)
+  return getRefs(doc)
+    .map(v => parseRef(v).url)
+    .filter(u => u !== '')
+    .map(u => path.resolve(path.join(dir, u)))
+}
+
+export class Graph {
+  readonly graph: Map<string, string[]>
```
Comment (sergey-shandar) [inline src/graph.ts]: Instead of `string[]` use `Document`. Where `Document` is `{ readonly references: () => Iterable<string> }`

Comment (sergey-shandar) [reply src/graph.ts]: Another option to define the `Document` is 
```ts
type Document = {
   readonly filePath: string
   readonly references: Iterable<Document> 
}
```

File: src/graph.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    let edges: string[] = this.adjacent(nodeA)
+    edges.push(nodeB)
+
+    edges = this.adjacent(nodeB)
+    edges.push(nodeA)
+  };
+
+  adjacent(node: string): string[] {
+    return this.graph.get(node) === undefined ? [] : this.graph.get(node) as string[]
```
Comment (sergey-shandar) [inline src/graph.ts]:
```ts
const result = this.graph.get(node)
return result === undefined ? [] : result
```

File: src/graph.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  buildFromDir(dirPath: string) {
+    let queue: string[] = []
```
Comment (sergey-shandar) [inline src/graph.ts]: use recursion instead of `queue`

File: src/graph.ts (hunk 1 above)
Comment (sergey-shandar) [reply src/graph.ts]: Targets:
1. make sure there is no circular references. If there are circular references, break a cycle and report an error.
2. `const getAllReferencedDocuments = (document: Document) => Iterable<Document>`. Remove duplication `it.uniq`.

File: src/graph.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+const getReferencedFileNames = (fileName: string, doc: json.Json) => {
+  const dir = path.dirname(fileName)
+  return getRefs(doc)
+    .map(v => parseRef(v).url)
+    .filter(u => u !== '')
+    .map(u => path.resolve(path.join(dir, u)))
+}
+
+export class Graph {
```
Comment (sergey-shandar) [inline src/graph.ts]: https://github.com/ts-common/ts-common.github.io#recommendations

File: src/graph.ts (hunk 2 above)
Comment (ruowan) [reply src/graph.ts]: I have deleted this file

File: src/graph.ts (hunk 3 above)
Comment (ruowan) [reply src/graph.ts]: done. use DFS recursion traversal

**Comment by ruowan** [comment]
It's ready to review.

File: src/index.ts [hunk 5]
```diff
@@ -117,53 +123,80 @@ const getReferencedFileNames = (fileName: string, doc: json.Json) => {
...
+ * @param whiteSet files set which haven't been explored yet
+ * @param graySet files currently being explored
+ * @param blackSet files have been explored
+ * @param inputFileSet: all referenced file
  */
-const resolveFileReferences = (readMePath: string, fileNames: Set<string>) =>
+const DFSTraversalValidate = (
+  current: string,
+  readMePath: string,
+  whiteSet: Set<string>,
```
Comment (sergey-shandar) [inline src/index.ts]: whiteSet should be just an input, which is an iterator.

File: src/index.ts [hunk 6]
```diff
@@ -117,53 +123,80 @@ const getReferencedFileNames = (fileName: string, doc: json.Json) => {
...
+  graySet: Set<string>,
+  blackSet: Set<string>,
```
Comment (sergey-shandar) [inline src/index.ts]: graySet and blackSet are internal sets, we don't need them as parameters.

File: src/index.ts [hunk 7]
```diff
@@ -117,53 +123,80 @@ const getReferencedFileNames = (fileName: string, doc: json.Json) => {
...
+ * @param readMePath current `readme.md` path
+ * @param whiteSet files set which haven't been explored yet
+ * @param graySet files currently being explored
+ * @param blackSet files have been explored
+ * @param inputFileSet: all referenced file
  */
-const resolveFileReferences = (readMePath: string, fileNames: Set<string>) =>
+const DFSTraversalValidate = (
+  current: string,
+  readMePath: string,
```
Comment (sergey-shandar) [inline src/index.ts]: we would like to analyze the complete graph.

**Pull request merged**